src/Graphex/Cabal.hs

```diff
@@ -54,30 +64,78 @@ sourceDirToFilePath :: FilePath -> FilePath
   gpd <- readGenericPackageDescription silent cabalFile
   let PackageDescription{..} = flattenPackageDescription gpd
   let candidateModules = mconcat
         [ do
-            Library{..} <- maybeToList library
+            guard $ Set.member CabalLibraries toDiscover
+            Library{..} <- mconcat [maybeToList library, subLibraries]
             srcDir <- hsSourceDirs libBuildInfo
             exMod <- exposedModules
             pure Module
               { name = fromString $ mconcat $ intersperse "." $ Cabal.components exMod
-              , path = sourceDirToFilePath srcDir </> Cabal.toFilePath exMod <.> ".hs"
+              , path = ModuleFile $ sourceDirToFilePath srcDir </> Cabal.toFilePath exMod <.> ".hs"
+              }
+        , do
+            guard $ Set.member CabalExecutables toDiscover
+            Executable{..} <- executables
+            srcDir <- hsSourceDirs buildInfo
+            otherMod <- "Main" : buildInfo.otherModules
+            pure Module
+              { name = fromString $
+                if otherMod == "Main"
+                then unUnqualComponentName exeName ++ "-Main"
+                else mconcat $ intersperse "." $ Cabal.components otherMod
+              , path = ModuleFile $ sourceDirToFilePath srcDir </> Cabal.toFilePath otherMod <.> ".hs"
+              }
+        , do
+            guard $ Set.member CabalTests toDiscover
+            TestSuite{..} <- testSuites
+            srcDir <- hsSourceDirs testBuildInfo
+            otherMod <- testBuildInfo.otherModules
+            pure Module
+              { name = fromString $ mconcat $ intersperse "." $ Cabal.components otherMod
+              , path = ModuleFile $ sourceDirToFilePath srcDir </> Cabal.toFilePath otherMod <.> ".hs"
               }
-        ] -- TODO: exes + other-modules
+        ]
+
+  traverse validateModulePath candidateModules
+
+validateModulePath :: Module -> IO Module
+validateModulePath m = do
+  path <- case m.path of
+    ModuleFile fp -> do
+      fileExists <- doesFileExist fp
+      pure $ if fileExists then ModuleFile fp else ModuleNoFile
+    ModuleNoFile -> pure ModuleNoFile
+  pure Module {name = m.name, path = path}
+
+data CabalModuleType =
+  CabalLibraries
+  | CabalExecutables
+  | CabalTests
+  deriving stock (Show, Eq, Ord)
 
-  filterM (doesFileExist . (.path)) candidateModules
+data CabalDiscoverOpts = CabalDiscoverOpts
+  { toDiscover      :: Set CabalModuleType
+  , includeExternal :: Bool
+  } deriving stock (Show, Eq)
 
-discoverCabalModuleGraph :: IO ModuleGraph
-discoverCabalModuleGraph = do
+discoverCabalModuleGraph :: CabalDiscoverOpts -> IO ModuleGraph
+discoverCabalModuleGraph opts@CabalDiscoverOpts{..} = do
   fs <- getDirectoryContents "." -- XXX
-  mods <- fmap fold . traverse discoverCabalModules . filter ((".cabal" ==) . takeExtension) $ fs
+  mods <- fmap fold . traverse (discoverCabalModules opts) . filter ((".cabal" ==) . takeExtension) $ fs
```

Comment:
I keep feeling like `fmap fold . traverse` should be a thing.

Reply:
It's `foldMapA`? But it's one pass:

```haskell
foldMapA
  :: forall t f a b
   . Foldable t
  => Applicative f
  => Monoid b
  => (a -> f b)
  -> t a
  -> f b
foldMapA f = getAp . foldMap (Ap . f)
```

Hm I should probably write a proposal to add it to `base`. I always write it in new projects 😆